### ISSUE TITLE
Hooks on param AccumulateGrad are not called when the param was called with set_

### ISSUE THREAD

**Issue by jerrychenhf** [pytorch/pytorch#139742]
### 🐛 Describe the bug

Background of the issue:
DeepSpeed depends a lot on param.data = other.data for ZeRO3 parameter offload. And ZeRO3 also depends on register a hook on param AccumulateGrad object to do reducing of grad partitions.

While torch.compile is converting param.data = other.data to param.set_. With param.set_ being called, we found the hooks on the param AccumulateGrad will not be called thus causing the accuracy issues. (https://github.com/microsoft/DeepSpeed/pull/5590)

The following program is a simple reproducer on CPU for showing the problem:
1. When "param.set_(r)" is used, the reduce_partition_and_remove_grads registered as AccumulateGrad hook is not called.
2. When replacing "param.set_(r)" with "param.data = r.data", reduce_partition_and_remove_grads registered as AccumulateGrad hook will be called. (Please don't keep the param._set(r) statement, as long as param.set_(r) is called even if is followed by  "param.data = r.data", the hook will not be called.

```
import torch
import torch.nn as nn

@torch.no_grad()
def do_gather(param):
  # Simulating the deepspeed gather
  r = torch.ones(param.shape, dtype=param.dtype, device=param.device)
  # reduce_partition_and_remove_grads will not be not called if used
  param.set_(r)
  # reduce_partition_and_remove_grads will be called if used
  # param.data = r.data

class SomeModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.param = nn.Parameter(torch.randn(5, 5))

    def forward(self, x):
        do_gather(self.param)
        return self.param ** 2 * x


model = SomeModule()

input = torch.randn(5,5)

_grad_accs = []
_grad_acc_hooks = []

def wrapper(param):
    print("HAIFENG wrapper PARAM:", id(param))
    param_tmp = param.expand_as(param)
    grad_acc = param_tmp.grad_fn.next_functions[0][0]

    def reduce_partition_and_remove_grads(*notneeded):
        print("HAIFENG reduce_partition_and_remove_grads PARAM:", id(param))

    _grad_acc_hooks.append(grad_acc.register_hook(reduce_partition_and_remove_grads))
    # needs to record otherwise, will not be called?
    _grad_accs.append(grad_acc)


model.train()                   
model.zero_grad()

wrapper(model.param)

print("FROWARD")
output = model(input)
loss = output.sum()
print(loss)

print("BACKWARD")
loss.backward()
print("param GRAD", model.param.grad)

```

Is this a BUG as to the grad hook for set_?
How should we get the grad hook called if it is not a bug?

### Versions

Collecting environment information...
PyTorch version: 2.6.0.dev20240914+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-122-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        43 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               12
On-line CPU(s) list:                  0-11
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   1
Core(s) per socket:                   6
Socket(s):                            2
Stepping:                             0
BogoMIPS:                             4190.15
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xsaves arat pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    VMware
Virtualization type:                  full
L1d cache:                            384 KiB (12 instances)
L1i cache:                            384 KiB (12 instances)
L2 cache:                             12 MiB (12 instances)
L3 cache:                             55 MiB (2 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-11
Vulnerability Gather data sampling:   Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:          KVM: Mitigation: VMX disabled
Vulnerability L1tf:                   Mitigation; PTE Inversion; VMX flush not necessary, SMT disabled
Vulnerability Mds:                    Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; IBRS; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.1.1
[pip3] torch==2.6.0.dev20240914+cpu
[conda] Could not collect

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan

**Comment by janeyx99** [comment]
This doesn’t directly answer your question, but we now have a register_post_accumulate_grad_hook API that is more supported/less sketchy than registering hooks on the AccGrad node. Could you try that hook and see if this behavior still persists? 

Also, both set_ and .data assignment are not recommended today. A safer method for two tensors to share a data field is doing something like y = x.detach().

**Comment by jerrychenhf** [comment]
@janeyx99  Thank you very much for the information!

I tried "register_post_accumulate_grad_hook" on the param Tensor and it works perfectly for both param.set_ case and param.data assignment case.  So register_post_accumulate_grad_hook is the more robust and preferred way for the case to hook for reducing graduate after the grad is accumulated.

> Also, both set_ and .data assignment are not recommended today. A safer method for two tensors to share a data field is doing something like y = x.detach().

Because we need to update the module parameter in-place after the parameters are gathered from ranks, a set_ or .data assignment is unavoidable after x.detach, am I right for this understanding?

**Comment by janeyx99** [comment]
@jerrychenhf https://pytorch.org/tutorials/recipes/recipes/swap_tensors.html#torch-utils-swap-tensors If I understand correctly, swap_tensors may be what you’re looking for?

**Comment by jerrychenhf** [comment]
Thanks @janeyx99  for the information.
Since register_post_accumulate_grad_hook is the preferred and robust way to hook for grad accumulation events, willo close the issue.
